jolt-core/src/jolt/vm/instruction_lookups.rs

```diff
@@ -529,144 +527,191 @@ where
         polynomials: &InstructionPolynomials<F, G>,
         gamma: &F,
         tau: &F,
-    ) -> (
-        Vec<DensePolynomial<F>>,
-        Vec<DensePolynomial<F>>,
-        Vec<DensePolynomial<F>>,
-        Vec<DensePolynomial<F>>,
-    ) {
+    ) -> (Vec<DensePolynomial<F>>, Vec<DensePolynomial<F>>) {
         let gamma_squared = gamma.square();
-        let read_leaves = (0..Self::NUM_MEMORIES)
+
+        let read_write_leaves = (0..Self::NUM_MEMORIES)
             .into_par_iter()
-            .map(|memory_index| {
+            .flat_map_iter(|memory_index| {
                 let dim_index = Self::memory_to_dimension_index(memory_index);
 
-                let leaf_fingerprints = (0..self.num_lookups)
-                    // .into_par_iter()
+                let read_fingerprints: Vec<F> = (0..self.num_lookups)
                     .map(|i| {
                         let a = &polynomials.dim[dim_index][i];
                         let v = &polynomials.E_polys[memory_index][i];
                         let t = &polynomials.read_cts[memory_index][i];
                         mul_0_1_optimized(t, &gamma_squared) + mul_0_1_optimized(v, gamma) + a - tau
```

Comment:
Not necessary in this PR but worth trying the arkworks F::mul_and_sum for at least 2 of these terms